jina/clients/__init__.py

```diff
@@ -18,16 +31,130 @@ def py_client(**kwargs) -> 'PyClient':
+        await py_client(host='192.168.1.100', port_expose=55555).index(input_fn, output_fn)
+
+    .. note::
+        to perform `index`, `search` or `train`, py_client needs to be awaited, as it is a coroutine
+
     """
     from ..parser import set_client_cli_parser
     from ..helper import get_parsed_args
     from .python import PyClient
     _, args, _ = get_parsed_args(kwargs, set_client_cli_parser(), 'Client')
     return PyClient(args)
+
+
+def py_client_runtime(mode, input_fn, output_fn, **kwargs) -> None:
+    """ This method allows callback execution of client process in main process
+
+    PyClient writes the response from servicer in a PAIR-BIND socket.
+    Main process reads from it using PAIR-CONNECT.
+
+    """
+    from zmq.error import Again
+    from ..parser import set_client_cli_parser
+    from ..helper import get_parsed_args
+    _, args, _ = get_parsed_args(kwargs, set_client_cli_parser(), 'Client')
+
+    # setting this manually to avoid exposing on CLI
+    args.runtime = 'process'
+
+    with JinaLogger(context='PyClientRuntime') as logger:
+        with CtrlZmqlet(args=args, logger=logger, is_bind=False, is_async=False, timeout=10) as zmqlet:
+            with PyClientRuntime(args, mode=mode, input_fn=input_fn, output_fn=output_fn,
+                                 address=zmqlet.address, **kwargs):
+                counter = 0
+                while True:
+                    try:
+                        msg = zmqlet.sock.recv()
+                        if msg == b'TERMINATE':
+                            # ideal way of exit, but PyClient socket might have closed before we recv it here
+                            logger.debug('received terminate message from the client response stream')
+                            zmqlet.sock.send_string('')
+                            break
+                        on_error = kwargs['on_error'] if 'on_error' in kwargs else pprint_routes
+                        on_always = kwargs['on_always'] if 'on_always' in kwargs else None
+                        grpc_response = RequestProto.FromString(msg)
+
+                        callback_exec(response=grpc_response, on_done=output_fn, on_error=on_error,
+                                      on_always=on_always, continue_on_error=args.continue_on_error,
+                                      logger=logger)
+                        zmqlet.sock.send_string('')
+
+                    except Again:
+                        # TODO(Deepankar): this can be handled better?
+                        if counter == 5:
+                            logger.warning('Waited for 5 seconds for zmq BIND socket before exiting')
+                            break
+                        logger.debug('PyClient\'s BIND socket is not open yet. waiting for some time!')
+                        time.sleep(1)
+                        counter += 1
+                        continue
+
+
+class PyClientRuntime(BasePea):
```

Comment:
could you call it PyClientPea? Like this it won't affect the naming that will come with the `BasePea` refactor

Reply:
I was hoping to change such Processes which don't have any `Pea` kind of logic to a different naming convention after the refactoring - e.g. - gateway-grpc, gateway-rest, remote. client etc

Reply:
Sure I will